### PR TITLE
ARROW-2451: [Python] Handle non-object arrays more efficiently in custom serializer.

### PR DESCRIPTION
**Pull request by robertnishihara**
**Before this PR**

```python
import numpy as np
import pyarrow as pa

x = np.array(10 ** 8 * [True, False])
%time serialized_obj = pa.serialize(x).to_buffer()  # 10.4s
%time new_x = pa.deserialize(serialized_obj)  # 8.94s

x = np.array([str(i) for i in range(10 ** 7)])
%time serialized_obj = pa.serialize(x).to_buffer()  # 2.24s
%time new_x = pa.deserialize(serialized_obj)  # 2.03s
```

**After this PR**

```python
import numpy as np
import pyarrow as pa

x = np.array(10 ** 8 * [True, False])
%time serialized_obj = pa.serialize(x).to_buffer()  # 117ms
%time new_x = pa.deserialize(serialized_obj)  # 265us

x = np.array([str(i) for i in range(10 ** 7)])
%time serialized_obj = pa.serialize(x).to_buffer()  # 174ms
%time new_x = pa.deserialize(serialized_obj)  # 13.2ms
```

cc @devin-petersohn @mitar @pcmoritz

### CODE REVIEW

File: python/pyarrow/serialization.py [hunk 1]
```diff
@@ -37,11 +37,22 @@
 # python_to_arrow.cc)
 
 def _serialize_numpy_array_list(obj):
-    return obj.tolist(), obj.dtype.str
+    if obj.dtype.str != '|O':
```
Comment (robertnishihara) [inline python/pyarrow/serialization.py]: I'm not sufficiently familiar with numpy dtypes to know if this `|O` is the only relevant special case.

File: python/pyarrow/tests/test_serialization.py [hunk 2]
```diff
@@ -315,9 +315,12 @@ def test_default_dict_serialization(large_buffer):
 
 def test_numpy_serialization(large_buffer):
     for t in ["bool", "int8", "uint8", "int16", "uint16", "int32",
-              "uint32", "float16", "float32", "float64"]:
+              "uint32", "float16", "float32", "float64", "<U1", "<U2", "<U3",
+              "<U4", "|S1", "|S2", "|S3", "|S4", "|O"]:
```
Comment (robertnishihara) [inline python/pyarrow/tests/test_serialization.py]: probably worth checking out https://github.com/dask/distributed/blob/master/distributed/protocol/tests/test_numpy.py#L30-L65

see https://issues.apache.org/jira/browse/ARROW-1596

**Review by pcmoritz — APPROVED** [review]
Great solution, feel free to merge!

**Comment by robertnishihara** [comment]
Also fixes https://issues.apache.org/jira/browse/ARROW-2264.

File: python/pyarrow/serialization.py (hunk 1 above)
Comment (mitar) [reply python/pyarrow/serialization.py]: I think you could check here for `obj.dtype.hasobject`: https://docs.scipy.org/doc/numpy/reference/generated/numpy.dtype.hasobject.html

**Pull request closed without merging**